pkg/daemon/writer.go

```diff
@@ -126,24 +127,42 @@ func (nw *clusterNodeWriter) Run(stop <-chan struct{}) {
 		case <-stop:
 			return
 		case msg := <-nw.writer:
-			r := implSetNodeAnnotations(nw.client, nw.lister, nw.nodeName, msg.annos)
+			r := implSetNodeAnnotations(nw.client, nw.lister, nw.nodeName, msg.annos, msg.annosToDelete)
 			msg.responseChannel <- r
 		}
 	}
 }
 
 // SetDone sets the state to Done.
-func (nw *clusterNodeWriter) SetDone(dcAnnotation string) error {
+func (nw *clusterNodeWriter) SetDone(state *stateAndConfigs) error {
 	annos := map[string]string{
 		constants.MachineConfigDaemonStateAnnotationKey: constants.MachineConfigDaemonStateDone,
-		constants.CurrentMachineConfigAnnotationKey:     dcAnnotation,
+		constants.CurrentMachineConfigAnnotationKey:     state.currentConfig.GetName(),
 		// clear out any Degraded/Unreconcilable reason
 		constants.MachineConfigDaemonReasonAnnotationKey: "",
 	}
+
```

Comment:
@yuqi-zhang I'd like to draw your attention to this part of the code for a re-review. When I was testing this in my integration PR, I was going about this in a slightly different and hackier way.

Reply:
Something I just wanted to point out for posterity: [This code](https://github.com/openshift/machine-config-operator/pull/3817#discussion_r1294068032) that you commented on was why this worked but didn't work the way I thought it did; hence the changes I made.